File: easy/104_MaximumDepthofBinaryTree.go
/**
 * 104. Maximum Depth of Binary Tree
 * Easy
 * Given the root of a binary tree, return its maximum depth.
 *
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	return getDepth(root, 0)
}

func getDepth(root *TreeNode, depth int) int {
	if root == nil {
		return depth
	}

	depth++

	newDepthLeft, newDepthRight := 0, 0
	if root.Left != nil {
		newDepthLeft = getDepth(root.Left, depth)
	}

	if root.Right != nil {
		newDepthRight = getDepth(root.Right, depth)
	}

	if newDepthLeft > depth {
		depth = newDepthLeft
	}
	if newDepthRight > depth {
		depth = newDepthRight
	}

	return depth
}

File: easy/104_MaximumDepthofBinaryTree.py
# 104. Maximum Depth of Binary Tree
# Easy
# https://leetcode.com/problems/maximum-depth-of-binary-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        q = []
        q.append([root, 1])
        res = 1
        while len(q) > 0:
            nodePair = q.pop(0)
            node = nodePair[0]
            level = nodePair[1]
            res = max(res, level)
            if node.left != None:
                q.append([node.left, level + 1])
            if node.right != None:
                q.append([node.right, level + 1])
        
        return res